src/services/codefixes/addMissingAsync.ts

```diff
@@ -0,0 +1,87 @@
+/* @internal */
+namespace ts.codefix {
+    type ContextualTrackChangesFunction = (cb: (changeTracker: textChanges.ChangeTracker) => void) => FileTextChanges[];
+    const fixId = "addMissingAsync";
+    const errorCodes = [
+        Diagnostics.Argument_of_type_0_is_not_assignable_to_parameter_of_type_1.code,
+        Diagnostics.Type_0_is_not_assignable_to_type_1.code,
+        Diagnostics.Type_0_is_not_comparable_to_type_1.code
+    ];
+
+    registerCodeFix({
+        fixIds: [fixId],
+        errorCodes,
+        getCodeActions: context => {
+            const { sourceFile, errorCode, cancellationToken, program, span } = context;
+            const diagnostic = find(program.getDiagnosticsProducingTypeChecker().getDiagnostics(sourceFile, cancellationToken), getIsMatchingAsyncError(span, errorCode));

```

Comment:
Can this code be shared with `isMissingAwaitError` in `addMissingAwait.ts`?  It seems like finding a diagnostic with particular related information might be a general utility.

Reply:
Not quite! In the other quick fix, we're checking if the diagnostic's span matches the input span, then returning the diagnostic if it matches and has the target related span code. In this quick fix, we don't do the filtering on related span at the matching step, because we're looking to `find` the span of the related span (so we can use that instead). So, yes, there's a common iteration bit, but to handle the bits variable between the two is essentially just a `some` with variable condition, onto which we then append an optional extra filter/map on the final result. Which seems like a bit too multifunctional for a single helper for my tastes.

But it should totally feel similar, since I started  writing this quick fix by copying the `addMissingAwait` quick fix, which is why the span-matching check is almost identical in structure.